### DISCUSSION TITLE
Separate Y-axes for two marks, scaling rectangles, or computing relative counts

### DISCUSSION THREAD

**Discussion by bbollen23** [uwdata/mosaic#606] [Q&A] [answered]
Hi there,

Working with vg plot has been amazing so far. I just have one question about the situation that I'm in.

I have a chart with two rectangle marks. Each are binned and display a count. The "grey" barchart has no selections applied. The "blue" barchart has a selection applied based on the slider.

Is there anyway to make it so that the "blue" barchart is scaled relatively? For instance, the "y" value of the blue chart would be "vg.count()/vg.sum()" , where "sum" is also calculated over the same selection. Attached is how it is in its current form.

<img width="277" alt="Screenshot 2024-11-22 at 5 00 25 PM" src="https://github.com/user-attachments/assets/e4ef5381-efb9-4920-84c4-80b47be55b6d">

There were a couple options I thought of:

1. See if there was a way to make the two marks have separate Y-axes. I know this is true with two separate plots using vconcat, but we would like the marks to be overlayed on one another
2. A "scale" value that can be inputted into the mark to aritificially increase the y-value after its pull from the table
3. A sql query which calculates this on the fly.
4. Add a "group by" statement to the mark when querying the duckdb database (Using "vg.count" is not identical to "vg.sql`COUNT(*)::INTEGER`" since the latter will not automatically append a group by statement. If we could add the group by statement whenever it makes this database query, we could easily scale this based on sum as well).
5. Use Vconcat where the marks still overlay on one another (like absolute positioning of vconcat so they are technically two different plots (thus separate axes) but are in one "view").

A solution to any one of these would probably solve our use case. Any help is greatly appreciated. Below is kind of what we're looking for, but with the bottom chart overlayed on the top chart instead (so its not two separate charts)

<img width="285" alt="Screenshot 2024-11-22 at 5 18 07 PM" src="https://github.com/user-attachments/assets/30cfadd1-db27-46b2-852b-2875a6442811">

Thanks!

**Comment by declann** [comment]
+1!

I too want to compare distributions. It helps enormously if they are normalized. I think we are facing the same limitations.

I'm using density plots, which are [really useful](https://stulp.gmw.rug.nl/ggplotworkshop/comparingdistributions.html#density-plots) for comparing multiple distributions (see image below) - but they must be normalized for easier comparisons about shape.

I can't figure a way using either **layered** density plots or layered histograms to make normalized comparisons. Here are possible solutions I thought about:

- Add x/y `Normalized` [attribute](https://idl.uw.edu/mosaic/api/vgplot/attributes.html#y-scale-attributes) for density or both
- (same as 1. by @bbollen23 ) add x/y `Independent` attribute for independent scales (like [vega lite](https://vega.github.io/vega-lite/docs/resolve.html))

Examples using layers to show distribution comparisons (based on airlines example) - these work easily in [mosaic playground](https://mosaic-playground.netlify.app/) (just C&P)

- density marks: https://gist.github.com/declann/a25d2090d7a6e4df56d4ab93208d3053
- histogram/rectY marks: https://gist.github.com/declann/a4ebafad33080dca3a746fa06ab36e05

If I could just normalize these, I think it would be super practical for comparing distributions of different selections.

(the test would be being able to discover all the relationships in the [crossfilter example](https://idl.uw.edu/mosaic/examples/flights-200k.html) - but keeping the context of the overall distributions (because of the reference layer))

![unnamed-chunk-63-1](https://stulp.gmw.rug.nl/ggplotworkshop/bookdown-demo_files/figure-html/unnamed-chunk-63-1.png)

Thanks - and please post if you know a good workaround!

**Comment by jheer** [comment, marked as answer]
Having separate y-axes is not an option at this point. The underlying Plot library does not support dual-axis charts so there is no underlying mechanism for this.

That leaves scaling / normalization as an option. Unfortunately, there is not (at present) a clean way to dynamically normalize values relative to a selection. One option is to include a scalar subquery (as in the [normalized stock price example](https://idl.uw.edu/mosaic/examples/normalize.html)), but this won't incorporate the current selection criteria.

What we really want is to take the generated query, then normalize the results. We might either do this in browser or push it to the database (e.g., by taking the mark query and wrapping it within a larger query). But either way this will require additional transformation infrastructure. I'll think about how we might add this as a possible extension.

**Reply by jheer** [reply]
While not a complete solution (it won't help with histograms, for example), it would be pretty simple to add a `normalize` option to the `densityX` and `densityY` marks if that would be useful.

**Reply by declann** [reply]
Thanks for the response Jeff. For my usecase a `normalize` option for density plots would work. Happy to help if I can!

**Reply by jheer** [reply]
Great, a density 1D `normalize` option is added by #612.

**Reply by declann** [reply]
Nice one Jeff! It works, heres an example: https://github.com/declann/mosaic/blob/jh/density-normalize/specs/yaml/distribution-comparison.yaml

That I host here (select Distribution Comparison): https://67466f086dfff462422a010e--resplendent-taffy-94006a.netlify.app/

Thanks a lot!

**Comment by bbollen23** [comment]
Thanks for the response, @jheer ! We actually did try the subquery example but in our case, the issue is that we wanted aggregate functions like count and sum. But we couldn't append the group by statement manually. Is there any possible support for that? The solution we decided on still works well for our use case, however.
